README.md
# ttml2ssa
Convert TTML/XML/DFXP/VTT/SRT subtitles used by Netflix, HBO, Disney+, Prime Video and others to SRT, SSA/ASS or VTT format.

Note: `ttml2ssa` is *not* a full-featured TTML-to-SRT converter and only works on a small subset of TTML documents. Namely, documents that follow the formats seen on the aforementioned streaming services.

## Usage
```
positional arguments:
  input-files           subtitle files

optional arguments:
  -h, --help            show this help message and exit
  -o [output file], --output [output file]
                        output file with extension srt, ssa or ass
  -v, --version         displays the version of this application and exits
  --shift [ms]          increases all timestamps by the specified value. You
                        can use a negative number
  --scale-factor [number or label]
                        multiplies the timestamps by this mumber. You can use
                        also any of these labels: NTSC2PAL (23.976/25),
                        PAL2NTSC (25/23.976), NTSC2FILM (23.976/24), PAL2FILM
                        (25/24), FILM2NTSC (24/23.976), FILM2PAL (24/25)
  -l [language code], --lang [language code]
                        subtitle language code ('en', 'es', etc,). It's used
                        by the language filter
  -a [number], --video-aspect [number]
                        the aspect ratio of the video. It's used to calculate
                        the correct PlayResY and PlayResY options for the SSA
                        style. Default: 16/9
  --no-cosmetic-fix     disables a filter which makes some cosmetic changes,
                        like adding a space after the symbol '-' and the next
                        word
  --no-language-fix     disables a filter which can fix some wrong characters
                        in some specific languages
  --no-fix-amazon-errors
                        don't try to fix errors that may occur in subtitles
                        from Amazon
  -c [encoding], --charset [encoding]
                        the encoding of the input file
  --output-format [srt, ssa or vtt]
                        output format to use if an output file has not been
                        set
  -srt, --srt           equivalent to --output_format srt
  -vtt, --vtt           equivalent to --output_format vtt
  --no-italics          removes the italic tags from the dialog texts
  --no-top              all dialog will be displayed at the bottom of the
                        screen
  --ssa-fontname [font], -f [font]
                        the font name (default: arial)
  --ssa-fontsize [number], -fs [number]
                        the font size (default: 50)
  --ssa-primary-color [color], -pc [color]
                        the primary color in format AABBGGRR or color name
                        (default: white)
  --ssa-back-color [color], -bc [color]
                        the back color in format AABBGGRR or color name
                        (default: 40000000)
  --ssa-outline-color [color], -oc [color]
                        the outline color in format AABBGGRR or color name
                        (default: black)
  --ssa-bold, -b        the font will be in bold
  --ssa-italic, -i      the font will be in italic
  --no-timestamp-manipulation
                        no changes will be made on timestamps
  --no-fix-collisions   collisions on timestamps won't be fixed
  --no-remove-duplicated
                        duplicated texts won't be removed
  --min-sep-ms [ms]     minimum separation (in ms) between framestamps
                        (SSA/ASS output only)
```
If multiple subtitle files are supplied, the application will create the
output files using the input filenames, replacing the extension with srt or ssa.
Be carefull, if a file with the same output name already exists the application
will overwrite it without asking.

If an output filename is specified with the -o option, then only the first of
the input files will be converted, using the output filename.


### Common use cases

Simple conversion:
```
ttml2ssa subtitle_from_netflix.xml -o subtitle.ssa
```
or
```
ttml2ssa subtitle_from_netflix.xml -o subtitle.srt
```

Shift everything forward by 2 secs:
```
ttml2ssa --shift 2000 subtitle_from_netflix.xml -o subtitle.srt
```

Convert from one frame rate to another:
```
ttml2ssa --scale-factor 23.976/25 subtitle_from_netflix.xml -o subtitle.srt
```
or
```
ttml2ssa --scale-factor NTSC2PAL subtitle_from_netflix.xml -o subtitle.srt
```
Those examples will convert a subtitle made for a movie at 23.976 frames per second for a version sped up to 25 fps (very common in Europe).

Convert two subtitles:
```
ttml2ssa subtitle1.xml subtitle2.xml
```
That will convert the subtitles to subtitle1.ssa and subtitle2.ssa

Convert all subtitles in a folder:
```
ttml2ssa *.xml
```
All files with extersion xml will be converted to ssa.

### Library
ttml2ssa can also be used as a library for other applications.

Simple example:
```
from ttml2ssa import Ttml2Ssa

input_file = 'subtitle.xml'
ttml = Ttml2Ssa()
ttml.parse_subtitle_file(input_file)
ttml.write2file('output.srt') # Saves as SRT
ttml.write2file('output.ssa') # Saves as SSA
```
Without loading or writing files:
```
from ttml2ssa import Ttml2Ssa

ttml_document = 'THIS STRING SHOULD CONTAIN AN XML SUBTITLE DOCUMENT'
ttml = Ttml2Ssa()
ttml.parse_ttml_from_string(ttml_document)
result = ttml.generate_srt()
# or
result = ttml.generate_ssa()
```

### Addon for Kodi
There's also an addon for Kodi
([script.module.ttml2ssa](https://github.com/Paco8/ttml2ssa/releases)),
which other addons can use to convert subtitles to SRT or SSA.
It provides a configuration dialog where the user can configure the SSA style and other stuff.

You can use it with something like this:
```
from ttml2ssa import Ttml2SsaAddon
...

ttml = Ttml2SsaAddon()
ttml.parse_subtitle_file(input_file)
# or
ttml.parse_ttml_from_string(subtitle_in_a_string)

ttml.write2file('output.srt')
# or
ttml.write2file('output.ssa')

# or
result = ttml.generate_srt()
# or
result = ttml.generate_ssa()
```

To open the ttml2ssa configuration dialog, you can add this to your addon settings.xml:
```
<setting label="SSA Settings" type="action" id="ssa_settings" option="close" action="Addon.OpenSettings(script.module.ttml2ssa)"/>
```

### Authors
ttml2ssa by Paco8, based on [ttml2srt](https://github.com/yuppity/ttml2srt) by yuppity.
License: LGPL-2.1


Changelog.md
# Changelog

## [0.3.8] - 2022-12-07
 * Remove the `<c></c>` tags from vtt files.
 * Duplicated entries are removed by default.
   The option `--no-remove-duplicated` disables this.
 * Add missing fields in ssa output. Fixes bug #6.

## [0.3.7] - 2021-12-04
 * Fix end of line when saving on Windows.

## [0.3.6] - 2021-11-15
 * Support timestamps without hour in vtt files.

## [0.3.5] - 2021-09-24
 * Fix parsing frameRateMultiplier.
 * Set the left and right margin to 0.

## [0.3.4] - 2021-03-20
 * Fix for Arabic in Amazon subtitles.
 * Remove the `<c></c>` tags from srt and vtt output.


src/timestampconverter.py
# encoding: utf-8
#
#  --------------------------------------------
#  based on https://github.com/yuppity/ttml2srt
#  --------------------------------------------
# SPDX-License-Identifier: GPL-2.0-or-later

from __future__ import unicode_literals, absolute_import, division
import re

class TimestampConverter(object):

    def __init__(self, frame_rate=23.976, tick_rate=1):
        self.tick_rate = tick_rate
        self.frame_rate = frame_rate

    def timeexpr_to_ms(self, *args):
        return self._timeexpr_to_ms(*args)

    def _timeexpr_to_ms(self, time_expr):
        """Use the given time expression to get a matching conversion method
        to overwrite self.timeexpr_to_ms() with.
        """

        self.timeexpr_to_ms = self.determine_ms_convfn(time_expr)
        return self.timeexpr_to_ms(time_expr)

    def _hhmmss_to_ms(self, hh, mm, ss):
        return hh * 3600 * 1000 + mm * 60 * 1000 + ss * 1000

    def subrip_to_ms(self, timestamp):
        """Desconstruct SubRip timecode down to milliseconds
        """

        hh, mm, ss, ms = re.split(r'[:,]', timestamp)
        return int(int(hh) * 3.6e6 + int(mm) * 60000 + int(ss) * 1000 + int(ms))

    def _metric_to_ms(self, metric_multiplier, metric_value):
        return int(metric_multiplier * metric_value)

    def _ms_to_hhmmssms(self, ms):
        hh = int(ms / 3.6e6)
        mm = int((ms % 3.6e6) / 60000)
        ss = int((ms % 60000) / 1000)
        ms = int(ms % 1000)

        return hh, mm, ss, ms

    def ms_to_subrip(self, ms):
        """Build SubRip timecode from milliseconds
        """

        hh, mm, ss, ms = self._ms_to_hhmmssms(ms)
        return '{:02d}:{:02d}:{:02d},{:03d}'.format(hh, mm, ss, ms)

    def ms_to_ssa(self, ms):
        """Build SSA/ASS timecode from milliseconds
        """

        hh, mm, ss, ms = self._ms_to_hhmmssms(ms)
        return '{:01d}:{:02d}:{:02d}.{:02d}'.format(hh, mm, ss, int(ms / 10))

    def frames_to_ms(self, frames):
        """Convert frame count to ms
        """

        return int(int(frames) * (1000 / self.frame_rate))

    def offset_frames_to_ms(self, time):
        """Convert offset-time expression with f metric to milliseconds.
        """

        frames = float(time[:-1])
        return int(int(frames) * (1000 / self.frame_rate))

    def offset_ticks_to_ms(self, time):
        """Convert offset-time expression with t metric to milliseconds.
        """

        ticks = int(time[:-1])
        seconds = 1.0 / self.tick_rate
        return (seconds * ticks) * 1000

    def offset_hours_to_ms(self, time):
        """Convert offset-time expression with h metric to milliseconds.
        """

        hours = float(time[:-1])
        return self._metric_to_ms(3.6e6, hours)

    def offset_minutes_to_ms(self, time):
        """Convert offset-time expression with m metric to milliseconds.
        """

        return self._metric_to_ms(60 * 1000, float(time[:-1]))

    def offset_seconds_to_ms(self, time):
        """Convert offset-time expression with s metric to milliseconds.
        """

        seconds = float(time[:-1])
        return self._metric_to_ms(1000, seconds)

    def offset_ms_to_ms(self, time):
        """Convert offset-time expression with ms metric to milliseconds.
        """

        ms = int(time[:-2])
        return ms

    def fraction_timestamp_to_ms(self, timestamp):
        """Convert hh:mm:ss.fraction to milliseconds
        """

        hh, mm, ss, fraction = re.split(r'[:.]', timestamp)
        hh, mm, ss = [int(i) for i in (hh, mm, ss)]
        # Resolution beyond ms is useless for our purposes
        ms = int(fraction[:3])

        return self._hhmmss_to_ms(hh, mm, ss) + ms

    def frame_timestamp_to_ms(self, timestamp):
        """Convert hh:mm:ss:frames to milliseconds

        Will handle hh:mm:ss:frames.sub-frames by discarding the sub-frame part
        """

        hh, mm, ss, frames = [int(i) for i in timestamp.split('.')[0].split(':')]
        hhmmss_ms = self._hhmmss_to_ms(hh, mm, ss)
        ms = self.frames_to_ms(frames)
        return hhmmss_ms + ms

    def determine_ms_convfn(self, time_expr):
        """Determine approriate ms conversion fn to pass the time expression to.

        Args:
            time_exrp (str): TTML time expression

        Return:
            Conversion method (callable)

        Strips the time expression of digits and uses the resulting string as
        a key to a dict of conversion methods.
        """

        # Map time expression delimiters to conversion methods. Saves
        # us from having to exec multibranch code on each line but assumes all
        # time expressions to be of the same form.
        time_expr_fns = {

            # clock-time, no frames or fraction
            # Example(s): "00:02:23"
            '::': self.frame_timestamp_to_ms,

            # clock-time, frames
            # Example(s): "00:02:23:12", "00:02:23:12.222"
            ':::': self.frame_timestamp_to_ms,
            ':::.': self.frame_timestamp_to_ms,

            # clock-time, fraction
            # Example(s): "00:02:23.283"
            '::.': self.fraction_timestamp_to_ms,

            # offset-time, hour metric
            # Example(s): "1h", "1.232837372637h"
            'h': self.offset_hours_to_ms,
            '.h': self.offset_hours_to_ms,

            # offset-time, minute metric
            # Example(s): "1m", "13.72986323m"
            'm': self.offset_minutes_to_ms,
            '.m': self.offset_minutes_to_ms,

            # offset-time, second metric
            # Example(s): "1s", "113.2312312s"
            's': self.offset_seconds_to_ms,
            '.s': self.offset_seconds_to_ms,

            # offset-time, millisecond metric
            # Example(s): "1ms", "1000.1231231231223ms"
            'ms': self.offset_ms_to_ms,
            '.ms': self.offset_ms_to_ms,

            # offset-time, frame metric
            # Example(s): "100f"
            'f': self.offset_frames_to_ms,
            '.f': self.offset_frames_to_ms,

            # offset-time, tick metric
            # Example(s): "19298323t"
            't': self.offset_ticks_to_ms,
            '.t': self.offset_ticks_to_ms,

        }

        try:
            delims = ''.join([i for i in time_expr if not i.isdigit()])
            return time_expr_fns[delims]
        except KeyError:
            raise NotImplementedError(
                'Unknown timestamp format ("{}")'.format(time_expr))


kodi/translations/create_pot.php
#!/usr/bin/php -q
# XBMC Media Center language file
# Addon Name: Ttml2Ssa
# Addon id: script.module.ttml2ssa
# Addon version: 0.1.1
# Addon Provider: Paco8
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: XBMC-Addons\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2020-12-28 16:47+0100\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE\n"
"Language: en\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"

<?php
function print_section($l, $n) {
    foreach ($l as $i) {
        echo "msgctxt \"#$n\"\n";
        echo "msgid \"$i\"\n";
        echo "msgstr \"\"\n\n";
        $n++;
    }
}

$style =  array('SSA Style', 
            'Font name', 'Font size', 
            'Primary colour (AABBGGRR or colour name)',
            'Back colour (AABBGGRR or colour name)',
            'Outline colour (AABBGGRR or colour name)',
            'Border style',
            'Outline with shadow', 'Opaque box',
            'Outline', 'Shadow',
            'Bold', 'Italic',
            'Vertical margin', 'Left margin', 'Right margin',
            'Main');

$misc = array('Misc',
            'Allow italics', 'Allow text on the top',
            'Allow cosmetic changes', 'Allow fixes for some languages',
            'Timestamp minimum separation (SSA only) in ms',
            'Allow timestamp manipulation', 'Fix collisions',
            'Try to fix Amazon errors',
            'Cache downloaded subtitles');

$addons = array('Subtitle type',
            'Improved Subtitles Settings',
            'Improved subtitles',
            'Normal (SRT)', 'Improved (SSA)', 'Both');

print_section($style, 32000);
print_section($misc, 32100);
print_section($addons, 32200);
?>


src/__main__.py
# encoding: utf-8
#
#  --------------------------------------------
#  based on https://github.com/yuppity/ttml2srt
#  --------------------------------------------
# SPDX-License-Identifier: LGPL-2.1-or-later

from __future__ import unicode_literals, absolute_import, division
from ttml2ssa import Ttml2Ssa

if __name__ == '__main__':

    import argparse

    argparser = argparse.ArgumentParser(
        description='Convert TTML/XML/DFXP/SRT/VTT subtitles to SubRip (SRT), SSA/ASS or WEBVTT (VTT) format.')
    argparser.add_argument('input-files',
        nargs="*",
        help='subtitle files',
        action='store')
    argparser.add_argument('-o', '--output',
        dest='output-file', metavar="output file",
        nargs='?',
        help='output file with extension srt, ssa or ass',
        action='store')
    argparser.add_argument('-v', '--version',
        dest="version",
        help="displays the version of this application and exits",
        action='store_true')
    argparser.add_argument('--shift',
        dest='shift',
        help='increases all timestamps by the specified value. You can use a negative number',
        metavar='ms', nargs='?',
        const=0, default=0, type=float,
        action='store')
    """
    argparser.add_argument('-f', '--fps',
        dest='sfps', metavar='fps',
        help='frames per second (default: 23.976)',
        nargs='?', const=23.976,
        default=23.976, type=float,
        action='store')
    """
    argparser.add_argument('--scale-factor',
        dest='scale', metavar='number or label',
        help='multiplies the timestamps by this mumber. You can use also any of these labels: '
             'NTSC2PAL (23.976/25), PAL2NTSC (25/23.976), '
             'NTSC2FILM (23.976/24), PAL2FILM (25/24), '
             'FILM2NTSC (24/23.976), FILM2PAL (24/25)',
        nargs='?', type=str, default='1', action='store')
    argparser.add_argument('-l', '--lang',
        dest="lang", metavar="language code",
        help="subtitle language code ('en', 'es', etc,). It's used by the language filter",
        nargs="?", type=str, default=None, action='store')
    argparser.add_argument('-a', '--video-aspect',
        dest='aspect', metavar='number',
        help='the aspect ratio of the video. '
             'It\'s used to calculate the correct PlayResY and PlayResY options for the SSA style. '
             'Default: 16/9',
        nargs='?', type=str, default='16/9', action='store')
    argparser.add_argument('--no-cosmetic-fix',
        dest="cosmetic_fix",
        help="disables a filter which makes some cosmetic changes, like adding a space after the symbol '-' and the next word",
        action='store_false')
    argparser.add_argument('--no-language-fix',
        dest="language_fix",
        help="disables a filter which can fix some wrong characters in some specific languages",
        action='store_false')
    argparser.add_argument('--no-fix-amazon-errors',
        dest="fix_amazon",
        help="don't try to fix errors that may occur in subtitles from Amazon",
        action='store_false')
    argparser.add_argument('-c', '--charset',
        dest='encoding', metavar='encoding',
        help='the encoding of the input file',
        nargs='?',
        default=None, type=str,
        action='store')
    argparser.add_argument('--output-format',
        dest='output_format', metavar="srt, ssa or vtt",
        nargs='?',
        help='output format to use if an output file has not been set',
        default='ssa', choices=['srt', 'ssa', 'vtt'],
        action='store')
    argparser.add_argument('-srt', '--srt',
        dest='output_format',
        help='equivalent to --output_format srt',
        const='srt',
        action='store_const')
    argparser.add_argument('-vtt', '--vtt',
        dest='output_format',
        help='equivalent to --output_format vtt',
        const='vtt',
        action='store_const')
    argparser.add_argument('--no-italics',
        dest='allow_italics',
        help='removes the italic tags from the dialog texts',
        action='store_false')
    argparser.add_argument('--no-top',
        dest='allow_top_pos',
        help='all dialog will be displayed at the bottom of the screen',
        action='store_false')

    argparser.add_argument('--ssa-fontname', '-f',
        dest='fontname', metavar='font',
        help='the font name (default: arial)',
        nargs='?',
        default='Arial', type=str,
        action='store')
    argparser.add_argument('--ssa-fontsize', '-fs',
        dest='fontsize', metavar='number',
        help='the font size (default: 50)',
        nargs='?',
        default=50, type=int,
        action='store')
    argparser.add_argument('--ssa-primary-color', '-pc',
        dest='primary_color', metavar='color',
        help='the primary color in format AABBGGRR or color name (default: white)',
        nargs='?',
        default='white', type=str,
        action='store')
    argparser.add_argument('--ssa-back-color', '-bc',
        dest='back_color', metavar='color',
        help='the back color in format AABBGGRR or color name (default: 40000000)',
        nargs='?',
        default='40000000', type=str,
        action='store')
    argparser.add_argument('--ssa-outline-color', '-oc',
        dest='outline_color', metavar='color',
        help='the outline color in format AABBGGRR or color name (default: black)',
        nargs='?',
        default='black', type=str,
        action='store')
    argparser.add_argument('--ssa-bold', '-b',
        dest='ssa_bold',
        help='the font will be in bold',
        action='store_true')
    argparser.add_argument('--ssa-italic', '-i',
        dest='ssa_italic',
        help='the font will be in italic',
        action='store_true')

    argparser.add_argument('--no-timestamp-manipulation',
        dest="timestamp_manipulation",
        help="no changes will be made on timestamps",
        action='store_false')
    argparser.add_argument('--no-fix-collisions',
        dest="fix_collisions",
        help="collisions on timestamps won't be fixed",
        action='store_false')
    argparser.add_argument('--no-remove-duplicated',
        dest="fix_duplicated",
        help="duplicated texts won't be removed",
        action='store_false')
    argparser.add_argument('--min-sep-ms',
        dest="ssa_timestamp_min_sep", metavar="ms",
        help="minimum separation (in ms) between framestamps (SSA/ASS output only)",
        nargs="?", type=int, default=200, action='store')

    args = argparser.parse_args()

    if args.scale in Ttml2Ssa.SCALE.keys():
        scale_factor = Ttml2Ssa.SCALE[args.scale]
    else:
        scale_factor = eval(args.scale)

    ttml = Ttml2Ssa(shift=args.shift, scale_factor=scale_factor, subtitle_language=args.lang)
    #ttml.source_fps = args.sfps
    ttml.use_cosmetic_filter = args.cosmetic_fix
    ttml.use_language_filter = args.language_fix
    ttml.fix_amazon_errors = args.fix_amazon
    ttml.allow_italics = args.allow_italics
    ttml.allow_top_pos = args.allow_top_pos
    ttml.allow_timestamp_manipulation = args.timestamp_manipulation
    ttml.fix_timestamp_collisions = args.fix_collisions
    ttml.fix_duplicated_entries = args.fix_duplicated
    ttml.ssa_timestamp_min_sep = args.ssa_timestamp_min_sep
    ttml.set_video_aspect_ratio(eval(args.aspect))

    ttml.ssa_style["Fontname"] = args.fontname
    ttml.ssa_style["Fontsize"] = args.fontsize
    ttml.ssa_style["PrimaryColour"] = ttml.string_to_color(args.primary_color)
    ttml.ssa_style["BackColour"] = ttml.string_to_color(args.back_color)
    ttml.ssa_style["OutlineColour"] = ttml.string_to_color(args.outline_color)
    ttml.ssa_style["Bold"] = -1 if args.ssa_bold else 0
    ttml.ssa_style["Italic"] = -1 if args.ssa_italic else 0

    if args.version:
        print("ttml2ssa version {}".format(ttml.VERSION))
        quit()

    input_files = getattr(args, 'input-files')
    output_file = getattr(args, 'output-file')

    if output_file and len(input_files) > 1:
        del input_files[1:]

    if not input_files:
        argparser.print_usage()

    for input_file in input_files:
        if output_file:
            output = output_file
        else:
            import os.path
            basename = os.path.splitext(input_file)[0]
            output = basename + "." + args.output_format
        print("Converting {} to {}".format(input_file, output))
        ttml.parse_subtitle_file(input_file, args.encoding)
        ttml.write2file(output)
